### DISCUSSION TITLE
How to properly extend/subclass `ConfigDict` to add additional options?

### DISCUSSION THREAD

**Discussion by cdwilson** [pydantic/pydantic#10419] [Question]
I'd like to extend `ConfigDict` to add some additional options for an ORM-like application.

To do this, I subclassed `ConfigDict` to add some additional options (e.g. allow the table name for the class to be configurable).

```python
from pydantic import BaseModel, ConfigDict

class OrmConfig(ConfigDict, total=False):
    table_name: str | None
```

Then, in the ORM's base model, I set `model_config` with some [global behavior](https://docs.pydantic.dev/latest/concepts/config/#change-behaviour-globally):

```python
class OrmModel(BaseModel):
    model_config: ClassVar[OrmConfig] = OrmConfig(  # type: ignore[reportIncompatibleVariableOverride]
        protected_namespaces=("orm_",),
    )
```

Finally, models can be defined with a custom `table_name`:

```python
class MyModel(OrmModel):
    model_config = {"table_name": "custom_table_name"}
```

This appears to be working, but couple questions:

1. Is it a bad idea to add custom options into Pydantic's `model_config` dict? I realize that there may be name conflicts with future [options](https://docs.pydantic.dev/latest/api/config/#pydantic.config.ConfigDict) added by Pydantic to `model_config`, but are there other "gotchas" that I might be missing by doing this? If there are reasons why this is a bad idea, or doesn't align with some planned changes to the way Pydantic's config system works, I'd just like to know now so I can look into other options.
2. Pylance does not like the fact that I'm overriding `model_config` in `OrmModel` (hence the `# type: ignore` above) and gives me the following error:
   ```
   "model_config" overrides symbol of same name in class "BaseModel"
     Variable is mutable so its type is invariant
       Override type "OrmConfig" is not the same as base type 
   "ConfigDict" Pylance(reportIncompatibleVariableOverride).
   ```
   I'm not super experienced with Python typing and was curious if there is a typing trick that I can do to make Pylance happy, with having to resort to `# type: ignore[reportIncompatibleVariableOverride]`
3. For people who have done this type of thing, would you recommend subclassing `ConfigDict` as I have described above or just create a totally separate config object that doesn't subclass `ConfigDict`? e.g.
   ```python
   class OrmConfig(TypedDict, total=False):
       table_name: str | None
   
   class OrmModel(BaseModel):
       orm_config: ClassVar[OrmConfig] = OrmConfig()
   ```
   This seems to have the benefits of preventing any naming conflicts, but then I have to add more logic in the metaclass's `__new__` to make `orm_config` behave similarly to `model_config` (e.g. [w/ respect to MRO](https://github.com/pydantic/pydantic/issues/9992))

**Comment by aryan-curiel** [comment]
I've been struggling with the same decision in a project I'm working on. I think I'll go with a separated config, jsut because my config is not strictly related to the model definition itself, and more into the use I'm going to give him.
I know other libraries use the `json_schema_extra` or however is called, but it feels weird to use it for anything not related to the json schema.

If anyone else has a better solution than any of these ones, please, post it here. I will be happy to read it.

**Comment by grabear** [comment]
+1.  

I'm curious what the pydantic devs recommend for extending pydantic model's `ConfigDict` or `SettingsConfigDict`.  I did something like this:
```python
from typing import ClassVar
from pydantic import BaseModel
from pydantic_settings import BaseSettings, SettingsConfigDict
from pydantic._internal._config import config_defaults
...

class ExtendedConfigDict(ConfigDict):
    allow_instantiation: bool
    """A flag that indicates whether the model is allowed instantiated directly.  This can be useful for base classes
    that don't function on their own."""


class ExtendedBaseModel(BaseModel):
    model_config: ClassVar[ExtendedConfigDict] = ExtendedConfigDict(
        allow_instantiation=True,
        **config_defaults,
    )

    def __new__(cls, *args, **kwargs):
        if "allow_instantiation" in cls.model_config:
            if not cls.model_config["allow_instantiation"]:
                raise TypeError(
                    f"The '{cls.__name__}' class cannot be instantiated.  "
                    f"Set 'allow_instantiation' to True in the model_config to allow instantiation."
                )
        return super().__new__(cls)
        
 class ExtendedSettingsConfigDict(SettingsConfigDict, ExtendedConfigDict):
    pass


class ExtendedBaseSettings(BaseSettings, ExtendedBaseModel):
    model_config: ClassVar[ExtendedSettingsConfigDict] = ExtendedSettingsConfigDict(
        allow_instantiation=True,
        **BaseSettings.model_config,
    )
```

...and it's been working for me... so far..

**Comment by grabear** [comment]
Any more thoughts here?
